kube-client/src/client/client_ext.rs

```diff
@@ -385,6 +395,54 @@ impl Client {
         req.extensions_mut().insert("list");
         self.request::<ObjectList<K>>(req).await
     }
+
+    /// Patch provided `Resource` implementing type `K`
+    ///
+    /// ```no_run
+    /// # use k8s_openapi::api::core::v1::Pod;
+    /// # use k8s_openapi::api::core::v1::Service;
+    /// # use kube::client::scope::{Namespace, Subresource};
+    /// # use kube::prelude::*;
+    /// # use kube::api::{PatchParams, Patch};
+    /// # async fn wrapper() -> Result<(), Box<dyn std::error::Error>> {
+    /// # let client: kube::Client = todo!();
+    /// let pod: Pod = client.get("some_pod", &Namespace::from("default")).await?;
+    /// let pp = &PatchParams::apply("controller").force();
+    /// // Perform an apply patch on the resource
+    /// client.patch(pod, pp, &Patch::Apply(&pod), &Default::default()).await?;
+    /// // Perform an apply patch on the resource status
+    /// client.patch(pod, pp, &Patch::Apply(&pod), &Subresource::Status).await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn patch<K, P: Serialize + Debug>(
+        &self,
+        resource: &K,
+        pp: &PatchParams,
+        patch: &Patch<P>,
+        subresource: &Subresource,
+    ) -> Result<K>
+    where
+        K: ResourceExt + Serialize + DeserializeOwned + Clone + Debug,
+        <K as Resource>::DynamicType: Default,
+    {
+        let name = resource.meta().name.as_ref().ok_or(Error::NameResolve)?;
+        let url = K::url_path(&Default::default(), resource.meta().namespace.as_deref());
+        let req = Request::new(url);
+        let req = match subresource {
+            Subresource::Core => req.patch(name, pp, patch).map_err(Error::BuildRequest)?,
+            Subresource::Status => req
+                .patch_subresource("status", name, pp, patch)
+                .map_err(Error::BuildRequest)?,
+            Subresource::Scale => req
+                .patch_subresource("scale", name, pp, patch)
+                .map_err(Error::BuildRequest)?,
+            Subresource::Custom(subresource) => req
+                .patch_subresource(subresource, name, pp, patch)
+                .map_err(Error::BuildRequest)?,
+        };
```

Comment:
interesting idea with the enum here. i am a little skeptical of it's benefit because it means the documentation must be smashed together for all the variants. separated is fairly easy to read on the client side, and doesn't require a new enum (that i'm pretty sure they'd only need to create for this call anyway).

the original style in api:

- `client.patch`
- `client.patch_status`
- `client.patch_scale`
- `client.patch_subresource`

feels preferable to me, unless you have any good reasons.